Release 07/05/2023

russtuck opened the pull request:
This Release will merge the following changes into production: 
- #934 
- #936 
- #941
This has been running on 360train since Monday 5:13pm.

EjPlatzer commented on Gordon360/Controllers/MembershipsController.cs:
```diff
@@ -83,6 +83,21 @@ public ActionResult<int> GetMembershipCount(string? activityCode = null, string?
             return Ok(result);
         }
 
+        /// <summary>
+        /// Gets the memberships of the given user
+        /// </summary>
+        /// <param name="username"> username filter </param>
+        /// <returns>The number of followers of the activity</returns>
+        [HttpGet]
+        [Route("student/{username}")]
+        [StateYourBusiness(operation = Operation.READ_ONE, resource = Resource.MEMBERSHIP)] // not sure why
+        public ActionResult<int> GetMembershipsOfUser(string username)
+        {
+            var result = _membershipService.GetMemberships(username: username, sessionCode: "*");
+
+            return Ok(result);
+        }
+
```
I don't understand why this route was added. All it does is implement d subset of the base `/memberships` route, in basically the same way as the `/profiles/{username}/memberships` route. The latter is better in terms of REST API design, but it was marked obsolete because the base `/memberships` route already covers all it's possible use cases with optional query params.

It looks like this was added for https://github.com/gordon-cs/gordon-360-ui/pull/1083, but the better route to use for that should be `/profiles/{username}/memberships-history`, the same way it's currently used for the Experience Transcript.

russtuck replied:
Antonia and Amos confirmed that the new endpoint is not needed, and in the process found an apparent security flaw in the recommended endpoint: if given no username, it returns all involvements for all users, without filtering to only public involvements.  Antonia is working on a fix for both issues.

EjPlatzer replied:
> Antonia and Amos confirmed that the new endpoint is not needed, and in the process found an apparent security flaw in the recommended endpoint: if given no username, it returns all involvements for all users, without filtering to only public involvements. Antonia is working on a fix for both issues.

If you're talking about the GetMembershipHistory method that I recommended (route `/profiles/{username}/memberships-history`), I'm fairly certain it can't be called without the `username` param. If you try, the API sends back this error:
```json
{
  "errors": {
    "username": [
      "The username field is required."
    ]
  },
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
  "title": "One or more validation errors occurred.",
  "status": 400,
  "traceId": "00-b582e79374a558d2ff3c822dc9ba1d2a-ddb6eacb502fbb4c-00"
}
```

russtuck replied:
Sorry, it's the base /memberships route which appears to have a security flaw.

EjPlatzer commented on Gordon360/Controllers/MembershipsController.cs:
```diff
@@ -218,7 +233,7 @@ public async Task<ActionResult<MembershipView>> SetGroupAdminAsync(int membershi
         /// <remarks>Calls the server to make a call and update the database with the changed information</remarks>
         /// <returns>The updated membership as a MembershipView object</returns>
-        [HttpPut]
+        [HttpPatch]
```
Why was this changed? This creates a breaking change for the UI, and it seems like it was included in an unrelated pull request?

amos-cha replied:
I believe this was a change to update the protocol after I added `patch` (since this is a purely updating route). I believe that this change is reflected in the UI PR.

EjPlatzer replied:
Technically, this should be a PUT route, not a PATCH route, since the resource we're acting on is `/memberships/{membershipID}/privacy`, that is the resource we're PUTting is the privacy field itself, so we aren't updating one specific part of the resource buet rather replacing the whole resource with the given value.

COnversely, if the route were simply `/memberships/{membershipID}`, but we were selectively updating some parts of the membership resource based on the parameters, that would be a proper PATCH method.

More importantly, the better way to handle breaking changes to the API is to mark the old route as obsolete (but leave it unchanged), and create the new route. Then, once all clients (in our case, the develop UI) has updated, we can remove the obsolete. That way, deploying the API will not break the UI. As it is currently, we would have to coordinate the upgrade of the API and UI so that they both update in tandem. That is not infeasible in our environment, although it can be a hassle because both have to wait on the other to be ready before merging. In larger-scale projects or APIs with multiple/third-party consumers, that can be the cause of a major outage.

russtuck commented on Gordon360/Controllers/MembershipsController.cs:
```diff
@@ -83,6 +83,21 @@ public ActionResult<int> GetMembershipCount(string? activityCode = null, string?
+        /// Gets the memberships of the given user
+        /// </summary>
+        /// <param name="username"> username filter </param>
+        /// <returns>The number of followers of the activity</returns>
```
If we don't delete this endpoint, then the comment will need improving.  The returns explanation looks wrong, and the summary should be more explicit about what's included: all memberships over all terms?  just public?  just those visible to the current user (so different for own profile than public profiles)?

russtuck commented:
We won't push this.  If the fixes can be done today, we'll test them for a day on train and consider pushing tomorrow.

russtuck commented:
Late removal of unused endpoint (never used in prod, and only briefly in train).

russtuck commented:
I think we've addressed all the issues.  But we don't want to merge until tomorrow morning, during working hours, to be sure it doesn't cause problems.
